#219 connection investments ptdf

DillonJ opened the pull request:
Hi folks,

So this PR implements PTDF friendly investments. The issue is documented in #216 but I will try to summarise here.

The formulation changes as follows:

DC load flow can be implemented using power transfer distribution factors (PTDFs). The flow on connection l using PTDFs is expressed as :

![image](https://user-images.githubusercontent.com/7080191/108827612-20406e80-75c6-11eb-8fee-82b657359260.png)
where PTDF(l,n) is the fraction of an injection at n that flows on l.

This is implemented in the constraint connection_flow_ptdf

Line outage distribution factors (LODF) express the fraction of the pre-outage flow on line(connection) k that flows on line l if line K is removed. In the context of investments, we can use LODFs to calculate the impact on network flows if investment candidate connection c is present or not.

To do this, we re-interpret the results of the above equation as the flows in the intact network, with all candidate connections invested in. We rename the constraint to `connection_intact_flow_ptdf` and we create a new variable that represents the intact network flows `connection_intact_flow`.

Then the actual flow on connection l is the `connection_intact_flow` plus the additional flow for every candidate connection not invested in. This is expressed in constraint connection_flow_intact_flow as follows:

![image](https://user-images.githubusercontent.com/7080191/108828552-49153380-75c7-11eb-80cc-e8faf63ff417.png). 

The term intact_flow(c) - flow(c) evaluates to intact_flow(c) if the connection is not invested in - otherwise the subsequent constraints ensure that intact_flow(c) - flow(c) is zero if connection c is invested in.

The following two constraints ensure that if connection c is invested in the flow on connection c is equal to intact_flow, otherwise it is zero. This is implemented in constraints candidate_connection_flow_intact_flow, connection_flow_capacity and candidate_connection_flow_intact_flow2 as follows:

![image](https://user-images.githubusercontent.com/7080191/108831999-34d33580-75cb-11eb-96a0-ea8d95630a4c.png)

This formulation avoids a multiplication of the investment variable and a flow variable.

DillonJ commented:
@manuelma perhaps you could take a look? I know you're busy though!

manuelma commented:
Will do! This one's exciting...

mihlema commented:
I think @iasonask might also have some background here.

manuelma commented:
> The term intact_flow(c) - flow(c) evaluates to intact_flow(c) if the connection is invested in - otherwise the subsequent constraints ensure that intact_flow(c) - flow(c) is zero if connection c is invested in.

Just to be sure, this should be ...The term intact_flow(c) - flow(c) evaluates to intact_flow(c) if the connection is **not** invested in...

manuelma commented:
flow(c) = {intact_flow(c) if c is invested in, zero otherwise} ?

DillonJ commented:
> > The term intact_flow(c) - flow(c) evaluates to intact_flow(c) if the connection is invested in - otherwise the subsequent constraints ensure that intact_flow(c) - flow(c) is zero if connection c is invested in.
> 
> Just to be sure, this should be ...The term intact_flow(c) - flow(c) evaluates to intact_flow(c) if the connection is **not** invested in...

You're right - I corrected the original comment

manuelma commented:
I would suggest

flow(c) > intact_flow(c) + (1 - connection_invested_available(c)) * capacity(c)

To avoid that 100000.

DillonJ commented:
> I would suggest
> 
> flow(c) > intact_flow(c) + (1 - connection_invested_available(c)) * capacity(c)
> 
> To avoid that 100000.

I considered that - but are we guaranteed there is a capacity defined for that flow?

manuelma commented:
> I considered that - but are we guaranteed there is a capacity defined for that flow?

With the 100.000, it results in intact_flow(c) < 100.000 when c not invested in... Isn't that dangerous? Maybe take the julia max float then?

manuelma commented on src/constraints/constraint_connection_flow_capacity_strict.jl:
```diff
@@ -0,0 +1,125 @@
+for the lowest resolution, and only the average of the higher resolution `connection_flow` is constrained.
+If instantaneous power needs to be constrained as well, defining the `connection_capacity` separately for each
+`connection_flow` can be used to achieve this.
+"""
```
Are new functions docstrings up to date? Not just copy/pasted? (Just making sure...)

DillonJ replied:
oops... they may have been copy/pasted lol - will fix them

manuelma commented on src/constraints/constraint_connection_flow_intact_flow.jl:
```diff
@@ -0,0 +1,131 @@
+#############################################################################
+
+"""
+    add_constraint_connection_flow_capacity!(m::Model)
```
This doesn't match the function name.

manuelma commented on src/constraints/constraint_connection_flow_intact_flow.jl:
```diff
@@ -0,0 +1,131 @@
+end
+
+"""
+    constraint_connection_flow_capacity_indices(m::Model; filtering_options...)
```
Doesn't match the function name.

manuelma commented on src/SpineOpt.jl:
```diff
@@ -128,7 +129,13 @@ include("constraints/constraint_node_state_capacity.jl")
-include("constraints/constraint_connection_flow_ptdf.jl")
+include("constraints/constraint_connection_flow_capacity_strict.jl")
+include("constraints/constraint_connection_intact_flow_capacity.jl")
+include("constraints/constraint_connection_flow_intact_flow.jl")
```
Maybe we can rename this to `constraint_connection_intact_flow` to avoid repetition of the word flow?

DillonJ replied:
I was trying to capture that the constraint is capturing the relationship between the two flows - perhaps `constraint_connection_flow_to_intact_flow` does that read better?

manuelma commented on src/SpineOpt.jl:
```diff
@@ -128,7 +129,13 @@ include("constraints/constraint_node_state_capacity.jl")
+include("constraints/constraint_connection_intact_flow_ptdf.jl")
+include("constraints/constraint_connection_intact_flow_ptdf_in_out.jl")
+include("constraints/constraint_candidate_connection_flow_intact_flow.jl")
+include("constraints/constraint_candidate_connection_flow_intact_flow2.jl")
```
Would you consider:

constraint_candidate_connection_flow_intact_flow -> constraint_candidate_connection_flow_lower_bound

constraint_candidate_connection_flow_intact_flow2 -> constraint_candidate_connection_flow_upper_bound

DillonJ replied:
that's nicer all right

manuelma commented on test/constraints/constraint_connection.jl:
```diff
@@ -820,5 +812,5 @@
         end
-    end
+    end    
 end
```
Thanks for fixing the tests. Would you consider writing tests for the new constraints? Maybe after we merge?

DillonJ replied:
Of course, that is the plan... I'm working on them

manuelma reviewed (changes requested):
I feel like reviewing a paper with this new system. I'm requesting a minor revision. In general, I like the approach and look forward to use in my case study.

DillonJ commented:
> > I considered that - but are we guaranteed there is a capacity defined for that flow?
> 
> With the 100.000, it results in intact_flow(c) < 100.000 when c not invested in... Isn't that dangerous? Maybe take the julia max float then?
> 
> **Edit**: `typemax(Int)` works inside `JuMP.@constraint`.

Actually, I updated pre-processing recently to copy the capacity "to the other side" when creating the additional relationships - so we are more than likely to have a capacity... we could just use the max int (or whatever) if there is no capacity defined I suppose

manuelma commented:
> I would suggest

> flow(c) > intact_flow(c) + (1 - connection_invested_available(c)) * capacity(c)

> To avoid that 100000.

> I considered that - but are we guaranteed there is a capacity defined for that flow?

So I suggest to use `typemax(Int)` instead.

If the flow capacity is not defined, what happens to the constraint just above? flow(c) <= connections_invested_available(c) * cap(c)?

Also, `connections_invested_available` can only be zero or one? If yes, why the plural?

DillonJ commented:
So @manuelma I believe I have addressed most of the issues. Now I use connection_capacity instead of 100000. If there is no capacity, the constraint falls back on 100000 but I made changes to pre-processing for bi-directional lossless lines to ensure that this should never be needed. I can't use typemax(Int) but it gets multiplied by the timeslice duration and we'd have an overflow otherwise.

manuelma commented on src/data_structure/preprocess_data_structure.jl:
```diff
@@ -177,6 +166,54 @@ function generate_direction()
+        for (n, d) in connection__from_node(connection=c)
+            if haskey(connection__from_node.parameter_values[(c, n, d)], :connection_capacity)
+                conn_capacity_param = connection__from_node.parameter_values[(c, n, d)][:connection_capacity]
+                found_from = true
```
`haskey()` before accessing the dictionary means two dict lookups. Would you consider using `get` instead? (it only requires one lookup)

```julia
found_value = get(dict, key, nothing)
if found_value != nothing
    # use found_value directly
else
    # not in dictionary
end

```

DillonJ replied:
ah, ok - good to know

manuelma replied:
I see that we look for connection_capacity only in parameter values, not in default values. Is that important?

We have the `indices` function that can collect keys where the value is different than `nothing`. It also accepts filters: 
```julia

for (c, n, d) in indices(connection_capacity; node=n, direction=d)
   ...
end
```

Maybe we can use it?

Also, I'm not a big fan of accessing and modifying the fields of `Parameter` values directly. SpineInterface should provide a function that does that. What do you think of something like `set_value!(parameter, class, index, value)`?

manuelma replied:
Did you add any tests for this code? If yes, I think we can merge and I can do a bit of refactoring later...

DillonJ replied:
> I see that we look for connection_capacity only in parameter values, not in default values. Is that important?
> 
> We have the `indices` function that can collect keys where the value is different than `nothing`. It also accepts filters:
> 
> ```julia
> for (c, n, d) in indices(connection_capacity; node=n, direction=d)
>    ...
> end
> ```
> 
> Maybe we can use it?
> 
> Also, I'm not a big fan of accessing and modifying the fields of `Parameter` values directly. SpineInterface should provide a function that does that. What do you think of something like `set_value!(parameter, class, index, value)`?

There are two issues here... if I use indices(connection_capacity, ...) I don't know if what I'm getting back is on `connection__from_node` or `connection__to_node` - the way I have written it is the only way I could figure out to know on which relationships and which indices exactly I need to set connection_capacity.

The other issue is that I have no idea what the parameter may look like - when I go to the dicts directly I can just copy the parameter_value and it doesn't matter if it's a static value or a timeseries or whatever... so if we were to add support in SpineInterface for this sort of thing - we would also need `get_value_raw(parameter, class, index, value)` (or whatever)?

DillonJ replied:
> Did you add any tests for this code? If yes, I think we can merge and I can do this bit of refactoring later...

You mean the pre-processing code? Not yet - but I have implemented tests for all the new constraints. I did test that the new pre-processing code works correctly, and I will write some specific tests for it. I don't have time today - but perhaps we could merge anyway and I will do it on Monday? I always get a little anxious of slipping behind and having to deal with merge conflicts so it would be nice to get it in if we could?

manuelma replied:
Let's merge and 'worry' about this later. You made fair points, and it looks like things need to get thought trough in SpineInterface. There's no point in delaying this PR because of that.

Everything else looks great, so I will finish my review so you can merge?

DillonJ commented:
Final test implemented - I believe there is now a test for everything I've added
